### PR TITLE
New persistence api module

### PR DESCRIPTION
**Pull request by dfakhritdinov**
This PR is first in the sequence of depended PRs:
1. [New persistence api module](https://github.com/evolution-gaming/akka-effect/pull/273), this one
2. [Interop w/ Akka Persistence plugins](https://github.com/evolution-gaming/akka-effect/pull/274)
3. [New persistent actor impl](https://github.com/evolution-gaming/akka-effect/pull/275)

The PR introduces new `persistence-api` module along with `SnapshotStore` and `EventStore` traits that describe persistence API for further implementation. The goal of new module is in providing persistence API to be implemented by 3rd party libraries like Kafka-Journal thus some traits were moved from `persistence` module into new one. All Akka related implementation stays _unchanged_ in `persistence` module.

### CODE REVIEW

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Event.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  def event: E
+  def seqNr: SeqNr
+
+}
+
+object Event {
+
+  private case class Const[E](event: E, seqNr: SeqNr) extends Event[E]
+
+  def const[E](event: E, seqNr: SeqNr): Event[E] = Const(event, seqNr)
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Event.scala:14]: could inside of `const` function

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedId.scala [hunk 2]
```diff
@@ -12,7 +12,10 @@ final case class EventSourcedId(value: String) {
 
 object EventSourcedId {
 
-  implicit val orderEventSourcedId: Order[EventSourcedId] = Order.by { a: EventSourcedId => a.value }
+  implicit val orderEventSourcedId: Order[EventSourcedId] = Order.by {
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedId.scala:15]: not sure we benefit from formatting here :)

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    * @tparam F
+    *   effect
+    * @tparam S
+    *   snapshot
+    * @tparam E
+    *   event
+    */
+  trait Recovery[F[_], S, E] {
+
+    def snapshot: Option[Snapshot[S]]
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala:53]: I think we should have `SnapshotStore.find` call (function name is up to you)

I am not in favour of this `Recovery` interface

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def events: Stream[F, Event[E]]
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala:54]: I think we should have `EventStore.find` call (function name is up to you)

I am not in favour of this `Recovery` interface

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshot.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  def snapshot: S
+  def metadata: Snapshot.Metadata
+
+}
+
+object Snapshot {
+
+  final case class Metadata(seqNr: SeqNr, timestamp: Instant)
+
+  private case class Const[S](snapshot: S, metadata: Snapshot.Metadata)
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshot.scala:16]: please keep things local / private with regards to scope

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshotter.scala [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    * Deletes all snapshots matching `criteria`.
+    *
+    * @return outer F[_] is about deletion in background, inner F[_] is about deletion being completed
+    */
+  def delete(criteria: Snapshotter.Criteria): F[F[Unit]]
+}
+
+object Snapshotter {
+
+  final case class Criteria(maxSequenceNr: Long = Long.MaxValue,
```
Comment (t3hnar) [inline persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshotter.scala:52]: formatting?

File: persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Events.scala [hunk 7]
```diff
@@ -1,144 +0,0 @@
...
-import cats.{Apply, Eval, NonEmptyTraverse, Order, Show}
-
-/**
-  *
-  * @param values inner Nel[A] will be persisted atomically,
-  *               This applies some restrictions on persistence layer.
-  *               Please do not overuse this feature
-  * @tparam A event
-  */
-final case class Events[+A](values: Nel[Nel[A]]) { self =>
```
Comment (t3hnar) [inline persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Events.scala:15]: sad that it does not recognise that it is just moved

**Review by t3hnar — CHANGES_REQUESTED** [review]

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshot.scala (hunk 5 above)
Comment (dfakhritdinov) [reply persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshot.scala]: fixed

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshotter.scala (hunk 6 above)
Comment (dfakhritdinov) [reply persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Snapshotter.scala]: I've fixed this case but bare in mind that I've added `scalafmt` due possibility of using it from VS Code. I will not reformat existing/unchanged files, but feel right to format any code that was moved from module to module

File: persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Events.scala (hunk 7 above)
Comment (dfakhritdinov) [reply persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/Events.scala:15]: well, yes, git not so smart. Thus full filed will be marked as created by me without any previous history

File: persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala (hunk 3 above)
Comment (dfakhritdinov) [reply persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala:53]: I introduced `Recovery` as logical step after loading snapshot and before starting streaming events: snapshot `seqNr` is starting point for event recovery, thus if both methods will be defined on top level of the stores then `EventStore.find` must take as param `seqNr`:

```
trait SnapshotStore[F[_], S] {
   def find: F[Option[Snapshot[S]]]
}

trait EventStore[F[_], E] {
   def stream(fromSeqNr: SeqNr): F[Stream[F, Event[E]]]
}
```
 
Also would it be reasonable to have two separate APIs for snapshots and events?

Comment (dfakhritdinov) [reply persistence-api/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedPersistence.scala]: done

**Review by t3hnar — APPROVED** [review]

**Pull request merged**